coderd/database/queries/proxies.sql

```diff
@@ -48,6 +49,38 @@ WHERE
 LIMIT
 	1;
 
+-- Finds a workspace proxy that has an access URL or app hostname that matches
+-- the provided hostname. This is to check if a hostname matches any workspace
+-- proxy.
+--
+-- The hostname must be sanitized to only contain [a-zA-Z0-9.-] before calling
+-- this query. The scheme, port and path should be stripped.
+--
+-- name: GetWorkspaceProxyByHostname :one
+SELECT
+	*
+FROM
+	workspace_proxies
+WHERE
+	-- Validate that the @hostname has been sanitized and is not empty. This
+	-- doesn't prevent SQL injection (already prevented by using prepared
+	-- queries), but it does prevent carefully crafted hostnames from matching
+	-- when they shouldn't.
+	--
+	-- Periods don't need to be escaped because they're not special characters
+	-- in SQL matches unlike regular expressions.
+	@hostname :: text SIMILAR TO '[a-zA-Z0-9.-]+' AND
```

Comment:
You can't do exact match because `app--agent--workspace--username.dev.coder.com` is not an exact match for `dev.coder.com` or `*.dev.coder.com`.

The specific lines you commented on is to avoid regex escapes in hostnames. It's a preventative measure to make the match more consistent and to avoid security issues.